kv/bulk: split/scatter after range says it is full

dt opened the pull request:
Previously we would split-and-scatter after adding "filling" a range so
that as we continued to fill, we would do so to a new range that was
first scattered to a new node. We did this once we had written 48MB to
one range, as determined by not hitting a split while sending that many
bytes during a single flush of the above buffer. 48MB was an OK default
when the default range size was 64MB, but has long since been far below
what we now consider full. This was a minor annoyance (causing merges)
but calling split-and-scatter so much more than is needed has recently
become a more severe issue now that ~every scatter call actually moves
a range and blocks on that snapshot.

Even when 48MB was a reasonable fraction of the 64MB default, this was
just a rough approximation for full that failed to account for the
actual range's configured size and only counted data flushed since the
last explicit Flush().

Instead, this change replaces this "split when i've sent enough" logic
with a new approach that lets the range itself reply to each SST with
its remaining capacity and end key. The caller can then use this to
determine if there is room to send another file to that range or if it
should consider it full, and split if/when another key that would be
added to that ranges span is added to the batch.

This significantly reduces how often ingestion stops to wait on scatter
of a split span with the recently updated scatter behavior, more than
doubling observed speeds of bulk ingestion.

Release justification: high impact fix to new/modified functionality.

Release note (performance improvement): Ranges are split and rebalanced during bulk ingestion only when they become full, reducing excessive splits and subsequent merging and associated rebalancing-related performance impact.

dt committed:
kv/bulk: split/scatter after range says it is full

Previously we would split-and-scatter after "filling" a range so that as
we continued to fill, we would do so to a new range that was first
scattered to a new node. We did this once we had written 48MB to one
range, as determined by not hitting a split while sending that many
bytes during a single flush of the above buffer. 48MB was an OK default
when the default range size was 64MB, but has long since been far below
what we now consider full. This was a minor annoyance (causing merges)
but calling split-and-scatter so much more than is needed has recently
become a more severe issue now that ~every scatter call actually moves a
range and blocks on that snapshot.

Even when 48MB was a reasonable fraction of the 64MB default, this was
just a rough approximation for full that failed to account for the
actual range's configured size and only counted data flushed since the
last explicit Flush().

Instead, this change replaces this "split when i've sent enough" logic
with a new approach that lets the range itself reply to each SST with
its remaining capacity and end key. The caller can then use this to
determine if there is room to send another file to that range or if it
should consider it full, and split if/when another key that would be
added to that ranges span is added to the batch.

This significantly reduces how often ingestion stops to wait on scatter
of a split span with the recently updated scatter behavior, more than
doubling observed speeds of bulk ingestion.

Release justification: high impact fix to new/modified functionality.

Release note (performance improvement): Ranges are split and rebalanced during bulk ingestion only when they become full, reducing excessive splits and subsequent merging and associated rebalancing-related performance impact.

dt committed:
kv/bulk: add a name to sst batcher for debug logging

Release justification: low-risk logging change.
Release note: none.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/77588)
<!-- Reviewable:end -->

dt committed:
kv/bulk: only log >0 commit-waits and track total wait

Release note: none.

nvb reviewed (commented):
> Reviewed 3 of 3 files at r1, 9 of 9 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77588)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @adityamaru, and @dt)

___
*[-- commits, line 10 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/77588#-MxoQRYN20p-7PDorAfk):*
`s/adding "filling"/"filling"/`
___
*[-- commits, line 18 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/77588#-MxoQeKvA3ol-Vbsb-Tf):*
Do you foresee issues with IMPORT in mixed version clusters? For instance, when a v21.2 node splits and scatters every 48MB while v22.1 actually respects those scatters? Is this issue severe enough to do something about it, like default all scatters issued by v21.2 nodes to no-ops?
___
*[pkg/kv/db.go, line 732 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/77588#-MxoRMRiC8yXfFJFRURU:-MxoRMRiC8yXfFJFRURV:b-8ajiqg) ([raw file](https://github.com/cockroachdb/cockroach/blob/bff1d47059a8cf7cc6c5f22827ef0be6d2cb9766/pkg/kv/db.go#L732)):*
> ```Go
> 	var span roachpb.Span
> 	var available int64
> 	for _, r := range b.response.Responses {
> ```

Is this loop needed? We're only sending a single request.

Same question down below.
___
*[pkg/kv/bulk/sst_batcher.go, line 296 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77588#-MxoQLuM4T8LN6s4qzwV:-MxoQLuM4T8LN6s4qzwW:b-q2oc0f) ([raw file](https://github.com/cockroachdb/cockroach/blob/ddfc8cf746b0f5705c7f1e0cce28d988579b48a0/pkg/kv/bulk/sst_batcher.go#L296)):*
> ```Go
> 			if r != nil {
> 				b.flushKey = r.Desc().EndKey.AsRawKey()
> 				log.VEventf(ctx, 3, "%s building sstable that will flush before %v", b.name, b.flushKey)
> ```

Should we use a context tag for this?
___
*[pkg/kv/bulk/sst_batcher.go, line 219 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/77588#-MxoSqyn7qWOKMAbjYHD:-MxoSqyn7qWOKMAbjYHE:b-g6mu0o) ([raw file](https://github.com/cockroachdb/cockroach/blob/bff1d47059a8cf7cc6c5f22827ef0be6d2cb9766/pkg/kv/bulk/sst_batcher.go#L219)):*
> ```Go
> 
> 	// If we're about to add more data to the same range that previously told us
> 	// its remaining capacity was less than the size of file we last sent it, we
> ```

Why do we consider the size of the file we last sent instead of the size of the file we're about to send? Does this make the code easier to reason about?
___
*[pkg/kv/bulk/sst_batcher.go, line 246 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/77588#-MxoTKwQ9v9QbWw51M2X:-MxoTKwQ9v9QbWw51M2Y:b-8r5l7o) ([raw file](https://github.com/cockroachdb/cockroach/blob/bff1d47059a8cf7cc6c5f22827ef0be6d2cb9766/pkg/kv/bulk/sst_batcher.go#L246)):*
> ```Go
> 					}
> 				}
> 				b.lastRangeSpan = reply.ScatteredSpan
> ```

Is this needed if we're about to set `b.lastRangeFull = false`?
___
*[pkg/roachpb/api.proto, line 1747 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/77588#-MxoTxIZ0ao9UALYehFi:-MxoTxI_42Gqfq0cCKDK:bjtp75n) ([raw file](https://github.com/cockroachdb/cockroach/blob/bff1d47059a8cf7cc6c5f22827ef0be6d2cb9766/pkg/roachpb/api.proto#L1747)):*
> ```protobuf
>   ResponseHeader header = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
> 
>   Span range_span = 2 [(gogoproto.nullable) = false];
> ```

Let's explain what these fields represent.


<!-- Sent from Reviewable.io -->

dt reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77588)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @adityamaru, @dt, and @nvanbenschoten)

___
*[-- commits, line 18 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/77588#-MxoQeKvA3ol-Vbsb-Tf):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do you foresee issues with IMPORT in mixed version clusters? For instance, when a v21.2 node splits and scatters every 48MB while v22.1 actually respects those scatters? Is this issue severe enough to do something about it, like default all scatters issued by v21.2 nodes to no-ops?
</blockquote></details>

I thought a little about how this change handles mixed-version, i.e. if it gets 0 available bytes replies from 21.2 kv, which is why we only use the available bytes reply if the span is valid. Otherwise we never decide the range is full and don't explicitly split, which seems fine based on the 448 experiment results.
A 21.2 importer talking to a 22.1 range that actually scatters will see the same "regression" -- that those scatters actually do work that takes time -- as master does right now. I don't know if it is worth any complexity to change the though.
___
*[pkg/kv/db.go, line 732 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/77588#-MxoRMRiC8yXfFJFRURU:-MxowtmiDUGaC0zVaILH:b8wryt2) ([raw file](https://github.com/cockroachdb/cockroach/blob/bff1d47059a8cf7cc6c5f22827ef0be6d2cb9766/pkg/kv/db.go#L732)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Is this loop needed? We're only sending a single request.

Same question down below.
</blockquote></details>

I suppose not. Removed.
___
*[pkg/kv/bulk/sst_batcher.go, line 296 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77588#-MxoQLuM4T8LN6s4qzwV:-Mxovbhv8yU9tcjxtXVH:b-7ro57d) ([raw file](https://github.com/cockroachdb/cockroach/blob/ddfc8cf746b0f5705c7f1e0cce28d988579b48a0/pkg/kv/bulk/sst_batcher.go#L296)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we use a context tag for this?
</blockquote></details>

IIRC we switched to putting explicit op names in various bulk vevents since context tags didn't reliably transit rpcs. That _might_ not be an issue here, if we locally control the context passed to the batcher/adder, though maybe it was also about the tags being captured in trace events? I don't remember exactly why. In any case, I'd rather change it wholesale separately than here.
___
*[pkg/kv/bulk/sst_batcher.go, line 219 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/77588#-MxoSqyn7qWOKMAbjYHD:-MxoVVY69S6kA8IJ_bP-:b-8t4ke9) ([raw file](https://github.com/cockroachdb/cockroach/blob/bff1d47059a8cf7cc6c5f22827ef0be6d2cb9766/pkg/kv/bulk/sst_batcher.go#L219)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Why do we consider the size of the file we last sent instead of the size of the file we're about to send? Does this make the code easier to reason about?
</blockquote></details>

We haven't actually made the file we're going to send next yet -- we just got the first key that will go in it. Prior to and after this change, before we proceed to make that file, we carve our a new range for it if we don't expect it to fit in the last one, either because we added "enough" (48mb) to the last one or because it replied that the remaining size was now less than if the next file is the same size.

We could move the split-and-scatter to after we finish a file and know how big it is, and go back and split at its first key if the range remaining size after the prior file is less than this new file. One drawback of that might be that if we split sooner, and get the actual split bounds back in the reply, it's less likely that while we make this next file the split bounds change?
___
*[pkg/kv/bulk/sst_batcher.go, line 246 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/77588#-MxoTKwQ9v9QbWw51M2X:-MxowTNV3AWzI7iaqp5s:b-sxtj5i) ([raw file](https://github.com/cockroachdb/cockroach/blob/bff1d47059a8cf7cc6c5f22827ef0be6d2cb9766/pkg/kv/bulk/sst_batcher.go#L246)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Is this needed if we're about to set `b.lastRangeFull = false`?
</blockquote></details>

Nope. Setting b.flushKey is probably useful though as it is more up-to-date than our range cache.


<!-- Sent from Reviewable.io -->

dt reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77588)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @adityamaru, @dt, and @nvanbenschoten)

___
*[pkg/kv/bulk/sst_batcher.go, line 219 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/77588#-MxoSqyn7qWOKMAbjYHD:-MxpKjKL43C3ZlJp9E_s:b-4ej6h0) ([raw file](https://github.com/cockroachdb/cockroach/blob/bff1d47059a8cf7cc6c5f22827ef0be6d2cb9766/pkg/kv/bulk/sst_batcher.go#L219)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

We haven't actually made the file we're going to send next yet -- we just got the first key that will go in it. Prior to and after this change, before we proceed to make that file, we carve our a new range for it if we don't expect it to fit in the last one, either because we added "enough" (48mb) to the last one or because it replied that the remaining size was now less than if the next file is the same size.

We could move the split-and-scatter to after we finish a file and know how big it is, and go back and split at its first key if the range remaining size after the prior file is less than this new file. One drawback of that might be that if we split sooner, and get the actual split bounds back in the reply, it's less likely that while we make this next file the split bounds change?
</blockquote></details>

I reworked this a bit to switch to wait until we're flushing a full file and see if it is too big to fit in the size reported at last flush, and then splits and scatters right before adding. Combined with the first-split case while I was at it.
___
*[pkg/roachpb/api.proto, line 1747 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/77588#-MxoTxIZ0ao9UALYehFi:-MxpKg9nBtnQdHGHazXB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bff1d47059a8cf7cc6c5f22827ef0be6d2cb9766/pkg/roachpb/api.proto#L1747)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Let's explain what these fields represent.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

dt commented:
Benchmark results, with the just-in-time split after we produce a file before we send it:
- imported 686 GiB bytes in 9 tables (took 47m19.491328384s, 247.34 MiB/s)
- imported 686 GiB bytes in 9 tables (took 45m37.823604944s, 256.52 MiB/s)
- imported 686 GiB bytes in 9 tables (took 44m36.128458896s, 262.44 MiB/s)

master:
- imported 686 GiB bytes in 9 tables (took 1h50m3.925839505s, 106.35 MiB/s)

dt commented:
@nvanbenschoten Anything else to address here?

dt commented:
Rebased on master to pick up @aayushshah15's scatter change. On a 10 node cluster (above numbers were 5 node I think, so just looking at master/fullsplits delta here), I see:

master: imported 686 GiB bytes in 9 tables (took... 
- `25m22s, 461.26 MiB/s`
- `26m05s,  448.63 MiB/s`
- `24m50s, 471.17 MiB/s`
- `25m53s, 452.00 MiB/s`.

fullsplits:
- `22m00s, 532.05 MiB/s`
- `22m16s, 525.66 MiB/s`
- `22m02s, 531.31 MiB/s`
- `22m15s, 525.98 MiB/s`
- `23m04s, 507.33 MiB/s`
- `22m36s, 517.82 MiB/s`

dt commented:
Wanted to get back to apples/apples numbers, so re-ran with 5node clusters like I had before:
master:
- `took 56m21.199067949s, 207.71 MiB/s`
- `took 55m0.655411006s, 212.78 MiB/s`

this branch:
- `took 42m58.443440139s, 272.38 MiB/s`
- `took 44m17.491131669s, 264.28 MiB/s`

So looks like still see some improvement over master, even with the fixed up scatter. And we won't expect the merge-storm an hour after IMPORT either.

dt commented:
Looks like we also see _significantly_ better compactions vs ingest rates with this
<img width="972" alt="Screen Shot 2022-03-16 at 11 31 25 AM" src="https://user-images.githubusercontent.com/15615/158627618-ece675f5-fe30-4d7b-80a2-d153aa1d5c9b.png">
<img width="967" alt="Screen Shot 2022-03-16 at 11 31 33 AM" src="https://user-images.githubusercontent.com/15615/158627621-565daf7b-36d6-4203-8586-9adbfa875b10.png">

dt committed:
kvserver,bulk: send ingest AdminScatter calls with a max-size condition

This adds a maximum size predicate to AdminScatterRequest to allow asking KV to scatter a
range if, and only if, its size does not exceed some limit. This can be used by mid-IMPORT
scatters, which are sent when the importing process had added data to the same range more
than once and that range is now 'full'; such scatters are generally good, and very cheap,
if they are sent to the freshly split off and empty RHS of a split that was created above
the previously ingested data in an in-order IMPORT, so that the next file can be sent to
the new range, scattered to a new node, and keep data well distributed without needing any
additional post-ingest rebalancing.

However this mis-ingest scatters can also go very wrong: if the RHS is _not_ empty, they are
no longer cheap; they have to actually relocate all the data in it. This can happen with
not-in-order IMPORTs. In a fully out of order import, such as a backfill, it is unlikely
that the ingestion process would want to split or scatter at all: as it generates data it
finds that after sorting a buffer, only a small amount of it is batched to any single range
and it is better off just letting the ranges decide when they are full and should split, and
then if they should be rebalanced. However if an IMPORT is overlapping or out of order at the
large scale, but in some local stretch of data appears mostly in-order, such that multiple SSTs
are sent to the same range, then the IMPORT process could be 'tricked' into trying to scatter
what it now assumes is an empty RHS. Since it is not actually empty, this is a bad idea, but it
does not know that, as it only sees its local stretch of keys, that it already read.

To fix this, this patch adds the maximum size argument to scatters and then lets IMPORT send
an AdminScatter that should be obeyed only if the range is mostly empty; a polite suggestion
or tip to KV that now _might be_ a good time to scatter that range before we keep filling it
if doing so is cheap.

Release note: none.

dt committed:
kv/bulk: create initial splits, then scatter

If the table is non-empty, e.g. in IMPORT INTO, splitting from left to
right and scattering it as we go means we scatter the last span k times,
since it is the suffix of the ever shrinking RHS.

Instead, split them all, then move each, now that it is only as wide as
up to the next split.

Release note: none.

dt commented:
@nvanbenschoten @aayushshah15 some additional changes added to this, specially the addition of a scatter-if-smaller-than-x argument on AdminScatter so that if we accidentally (e.g. during IMPORT INTO or with ordered but overlapping files) try to scatter full ranges, we don't accidentally go and move lots of data for no reason. I also added splitting _then_ scattering, so we don't scatter the last span k times when we split k ways

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, but someone more familiar with the bulk code here should also give this a pass.

Are you planning on trying to get this backported to release-22.1?

> Reviewed 1 of 11 files at r7, 12 of 12 files at r15, 2 of 2 files at r16, 10 of 10 files at r17, 6 of 6 files at r18, 2 of 2 files at r19, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77588#-:-MyjVey42oTqUsXvwshn:b-j8fj24)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aayushshah15, @adityamaru, and @dt)

___
*[pkg/kv/bulk/buffering_adder.go, line 59 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/77588#-MyjS2zu5CB34DMIuo2X:-MyjS2zu5CB34DMIuo2Y:bssu0oi) ([raw file](https://github.com/cockroachdb/cockroach/blob/a391949be8d9233b017ed3a63581c5fc5edbbe81/pkg/kv/bulk/buffering_adder.go#L59)):*
> ```Go
> 		totalFlush   time.Duration
> 		totalFilling time.Duration
> 		span         roachpb.Span
> ```

nit: does this belong in the `flishCounts` struct? Also, a comment on this field would be helpful.
___
*[pkg/kv/bulk/buffering_adder.go, line 33 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/77588#-MyjU947DeU1iZ9D9i3U:-MyjU947DeU1iZ9D9i3V:b-9p6an6) ([raw file](https://github.com/cockroachdb/cockroach/blob/6eac5629a1884f134f39deb480060e30a66a1dee/pkg/kv/bulk/buffering_adder.go#L33)):*
> ```Go
> )
> 
> const maxScatterSize = 4 << 20
> ```

nit: give this a bit of an explanation.
___
*[pkg/kv/bulk/buffering_adder.go, line 425 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/77588#-MyjUoig7GHUhvAitVVV:-MyjUoig7GHUhvAitVVW:b-r3lxv3) ([raw file](https://github.com/cockroachdb/cockroach/blob/6eac5629a1884f134f39deb480060e30a66a1dee/pkg/kv/bulk/buffering_adder.go#L425)):*
> ```Go
> 		b.sink.flushCounts.splitWait += timeutil.Since(beforeSplit)
> 		beforeScatter := timeutil.Now()
> 		resp, err := b.sink.db.AdminScatter(ctx, splitKey, 0)
> ```

nit: `0 /* maxSize */`
___
*[pkg/kv/bulk/sst_batcher.go, line 144 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/77588#-MyjSKm7-gUWpid59gIc:-MyjSKm7-gUWpid59gId:b71h34b) ([raw file](https://github.com/cockroachdb/cockroach/blob/a391949be8d9233b017ed3a63581c5fc5edbbe81/pkg/kv/bulk/sst_batcher.go#L144)):*
> ```Go
> 	flushKeyChecked bool
> 	flushKey        roachpb.Key
> 	// lastRangeSpan is used to tell if a send went to same range as prior send.
> ```

"as a prior send" or "as the prior send" or "as its prior send"
___
*[pkg/kv/bulk/sst_batcher.go, line 413 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/77588#-MyjVCt-0D8_mUo9rIpz:-MyjVCt-0D8_mUo9rIq-:bhexg7d) ([raw file](https://github.com/cockroachdb/cockroach/blob/6eac5629a1884f134f39deb480060e30a66a1dee/pkg/kv/bulk/sst_batcher.go#L413)):*
> ```Go
> 				b.flushCounts.scatterWait += timeutil.Since(beforeScatter)
> 				if err != nil {
> 					if strings.Contains(err.Error(), "existing range size") {
> ```

Let's leave a TODO about typing this error, like we have for the predicate key error.
___
*[pkg/roachpb/api.proto, line 1774 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/77588#-MyjTVOkEvYKCJgD2z_9:-MyjTVOl68ibq_Mf72-C:b-14wqwr) ([raw file](https://github.com/cockroachdb/cockroach/blob/a391949be8d9233b017ed3a63581c5fc5edbbe81/pkg/roachpb/api.proto#L1774)):*
> ```protobuf
>   Span range_span = 2 [(gogoproto.nullable) = false];
> 
>   // AvailableBytes indicates how many bytes remain before the range which into
> ```

"which into which"


<!-- Sent from Reviewable.io -->

dt commented:
> Are you planning on trying to get this backported to release-22.1?

I think we need to backport something to 22.1 or else we're looking at a pretty hefty regression in IMPORT performance with the new and more active AdminScatter that's already 22.1 (or back that out, but that'd make be sorta sad). It's _possible_ we could so something simpler for 22.1.0 than this, such as just disabling as-we-go split-and-scatters all together, but that seems risky for very large imports as we saw some slow increase in inbalance as we went. So for now, this is by best idea for a 22.1 fix, but eager to entertain simpler options if anyone has ideas.

dt reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77588)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aayushshah15, @adityamaru, @dt, and @nvanbenschoten)

___
*[pkg/kv/bulk/buffering_adder.go, line 59 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/77588#-MyjS2zu5CB34DMIuo2X:-MylYetoAIwc0h9FPx1S:bn61y4s) ([raw file](https://github.com/cockroachdb/cockroach/blob/a391949be8d9233b017ed3a63581c5fc5edbbe81/pkg/kv/bulk/buffering_adder.go#L59)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: does this belong in the `flishCounts` struct? Also, a comment on this field would be helpful.
</blockquote></details>

yeah, it's the total span _into which this adder has flushed_ and is just for debugging those flush stats. I want to pull this whole anon struct out and into a "IngestStats" proto or something (and unify with the one in the batcher) but I guess that should wait for a non-backporting change.
___
*[pkg/kv/bulk/buffering_adder.go, line 33 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/77588#-MyjU947DeU1iZ9D9i3U:-MymgPZVA-MlFIRo4PZj:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/6eac5629a1884f134f39deb480060e30a66a1dee/pkg/kv/bulk/buffering_adder.go#L33)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: give this a bit of an explanation.
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 6 of 10 files at r22, 4 of 6 files at r23, 2 of 2 files at r24, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77588#-:-Mymkkbw3eu12lXu4HLK:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aayushshah15, @adityamaru, and @dt)

___
*[pkg/kv/bulk/sst_batcher.go, line 150 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/77588#-Mymkbus5nfLYbmGE5q4:-Mymkbus5nfLYbmGE5q5:bp43iwg) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b41fabe7a5e2a030b2ff28959809c8c9b24fd6d/pkg/kv/bulk/sst_batcher.go#L150)):*
> ```Go
> 	flushKeyChecked bool
> 	flushKey        roachpb.Key
> 	// lastRange is span and remaining capacity of the last range added to, for
> ```

"is the span"


<!-- Sent from Reviewable.io -->

dt reviewed (commented):
TFTRs!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77588#-:-MymnZLP6K2klL7SV-ea:b-tlx7ko)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aayushshah15, @adityamaru, and @nvanbenschoten)

___
*[pkg/kv/bulk/sst_batcher.go, line 150 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/77588#-Mymkbus5nfLYbmGE5q4:-MymnY04Ed7qP-xzIW7b:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b41fabe7a5e2a030b2ff28959809c8c9b24fd6d/pkg/kv/bulk/sst_batcher.go#L150)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

"is the span"
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

dt commented:
bors r+

dt commented:
blathers backport 22.1

dt commented:
bors r+

dt commented:
blathers backport 22.1
